File: A4/code/lib-source/java/io/FileInputStream.java
/*
** simpleRTJ java language core library component.
**
** See the http://www.gnu.org/copyleft/gpl.html for
** details on usage and redistribution of this file.
*/

package java.io;

public class FileInputStream extends InputStream
{
    private int handle;
    int grand_total_read;

    public FileInputStream(String name) throws FileNotFoundException
    {
        this(new File(name));
    }

    public FileInputStream(File file) throws FileNotFoundException
    {
        if(!file.exists())
            throw new FileNotFoundException(file.getName());

        handle = open(file.getPath());
    }

    public int available() throws IOException
    {
        int len = getFileLength0(handle);
        if(len < 0)
            throw new IOException();

        len -= grand_total_read;
        if(len <= 0)
            return 0;

        return len;
    }

    public synchronized int skip(int num_bytes) throws IOException
    {
        if(num_bytes <= 0)
            return 0;

        if(num_bytes > available())
            num_bytes = available();

        int bytes_skipped = skip0(handle, num_bytes);
        if(bytes_skipped < 0)
            throw new IOException();

        grand_total_read += bytes_skipped;

        return bytes_skipped;
    }

    public synchronized int read() throws IOException
    {
        byte[] buf = new byte[1];

        int bytes_read = read0(handle, buf, 0, buf.length);
        if(bytes_read < 0)
            throw new IOException();

        if(bytes_read == 0)
            return -1;

        ++grand_total_read;

        return(buf[0] & 0xFF);
    }

    public int read(byte[] buf) throws IOException
    {
        return read(buf, 0, buf.length);
    }

    public synchronized int read(byte[] buf, int offset, int len) throws IOException
    {
        if(len == 0)
            return 0;

        int bytes_read = read0(handle, buf, offset, len);
        if(bytes_read < 0)
            throw new IOException();

        if(bytes_read == 0)
            return -1;

        grand_total_read += bytes_read;
        return bytes_read;
    }

    private int open(String name) throws FileNotFoundException
    {
        int rv = open0(name);
        if(rv < 0)
            throw new FileNotFoundException();

        return rv;
    }

    public synchronized void close() throws IOException
    {
        int rv = close0(handle);
        if(rv < 0)
            throw new IOException();

        handle = -1;
    }

    private native int close0(int handle);
    private native int open0(String name);
    private native int read0(int handle, byte[] buf, int offset, int len);
    private native int skip0(int handle, int num_bytes);
    private native int getFileLength0(int handle);
}

File: A4/code/lib-source/java/io/native/FileInputStream.c
/**************************************************************************************
*  FileInputStream.c
*
*  This file defines all native methods that have to be implemented to support
*  java/lang/FileInputStream class.
*  Actual implementation is target device dependent and therefore must be
*  implemented by the application developer.
*
*  Application developer should use the following functions to get access to
*  arrays and String object:
*       vmGetStringCount
*       vmGetStringArray
*       vmGetArraySize
*       vmGetArray
*  These functions are accessible via vm_config_t structure (see Win32 demo
*  application for examples on how to use them).
*
**************************************************************************************
*
*  The contents of this file are subject to the RTJ COMPUTING non-commercial
*  source code license agreement ("License"); you may not use this file except in
*  compliance with the License. A copy of the License (license.terms) can be
*  found in the root of installation folder or on the web at
*  http://www.rtjcom.com/Files/license.terms
*
*  Software distributed under the License is distributed on an "AS IS"
*  basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
*  License for the specific language governing rights and limitations
*  under the License.
*
*  Copyright (c) 2000 RTJ Computing. All right reserved.
*
*  Developer(s): Peter Gasparik <peterg@rtjcom.com>
**************************************************************************************/

#include "javavm.h"

/****************************************************************************
* Closes the opened file handle.
* On entry param[1] will contain native file handle.
* On return function should indicate if handle was closed via retval parameter:
*       0   - handle closed successfully
*       -1  - closing has failed
*
* Java prototype:
*           private native int close0(int handle);
*
****************************************************************************/
int16 java_io_FileInputStream_close0(int32 param[], int32 *retval)
{
    /* Add your code here */

    return NO_EXCEP;
}


/****************************************************************************
* Opens the named file.
* On entry param[1] will contain reference to String containg the name of a file.
* On return function should provide file handle via retval parameter:
*       >= 0    - file handle
*       -1      - opening has failed
*
* Java prototype:
*           private native int open0(String name);
*
****************************************************************************/
int16 java_io_FileInputStream_open0(int32 param[], int32 *retval)
{
    /* Add your code here */

    return NO_EXCEP;
}


/****************************************************************************
* Reads the data from the file.
* On entry param[1] will contain native file handle.
*          param[2] will contain reference to destination byte array
*          param[3] will contain offset to destination array
*          param[4] will contain number of bytes to read (copy to dest. array)
* On return function should provide number of bytes read via retval parameter:
*       >= 0    - actual number of bytes copied into destination array
*       -1      - reading has failed
*
* Java prototype:
*           private native int read0(int handle, byte[] buf, int offset, int len);
*
****************************************************************************/
int16 java_io_FileInputStream_read0(int32 param[], int32 *retval)
{
    /* Add your code here */

    return NO_EXCEP;
}


/****************************************************************************
* Skips the specified number of bytes.
* On entry param[1] will contain native file handle.
*          param[2] will contain number of bytes to skip
* On return function should provide number of bytes read via retval parameter:
*       >= 0    - actual number of bytes skipped
*       -1      - skipping has failed
*
* Java prototype:
*           private native int skip0(int handle, int num_bytes);
*
****************************************************************************/
int16 java_io_FileInputStream_skip0(int32 param[], int32 *retval)
{
    /* Add your code here */

    return NO_EXCEP;
}


/****************************************************************************
* Retrieves the file length.
* On entry param[1] will contain native file handle.
* On return function should provide file length via retval parameter:
*       >= 0    - length of the file
*       -1      - get file length failed
*
* Java prototype:
*           private native int getFileLength0(int handle);
*
****************************************************************************/
int16 java_io_FileInputStream_getFileLength0(int32 param[], int32 *retval)
{
    /* Add your code here */

    return NO_EXCEP;
}

